clash-prelude/src/Clash/Explicit/BlockRam/Model.hs

```diff
@@ -0,0 +1,311 @@
+import Clash.Sized.Index (Index)
+import Clash.XException (NFDataX(..), seqX)
+import Clash.XException.MaybeX (MaybeX(..), toMaybeX, andX)
+
+import Control.Applicative (liftA2)
+import Data.Maybe (fromMaybe)
+import Data.Sequence (Seq)
+import GHC.Stack (HasCallStack)
+import GHC.TypeNats (KnownNat)
+
+import qualified Data.Sequence as Seq
+import qualified Clash.XException.MaybeX as MaybeX
+
+-- | Helper used in 'getConflict'
+data Conflict = Conflict
+  { cfRWA :: !(MaybeX Bool) -- ^ Read/Write conflict for output A
+  , cfRWB :: !(MaybeX Bool) -- ^ Read/Write conflict for output B
+  , cfWW  :: !(MaybeX Bool) -- ^ Write/Write conflict
+  } deriving (Show)
+
+-- | Determines whether there was a write/write or read/write conflict. A conflict
+-- occurs when two ports tried to (potentially, in case of undefined values)
+-- access the same address and one or both tried to write to it. See documentation
+-- of 'Conflict' for more information.
+getConflict ::
+  -- | Port A: enable, write enable, address
+  (MaybeX Bool, MaybeX Bool, MaybeX Int) ->
+  -- | Port B: enable, write enable, address
+  (MaybeX Bool, MaybeX Bool, MaybeX Int) ->
+  -- | 'Just' if there is a (potential) write conflict, otherwise 'Nothing'
+  Maybe Conflict
+getConflict (enA, wenA, addrA) (enB, wenB, addrB)
+  | IsDefined False <- sameAddrX = Nothing
+  | otherwise                    = Just conflict
+ where
+  sameAddrX = liftA2 (==) addrA addrB
+
+  conflict = Conflict
+    { cfRWA = andX enA (andX enB wenB)
+    , cfRWB = andX enB (andX enA wenA)
+    , cfWW  = andX (andX enA enB) (andX wenA wenB)
+    }
+
+-- | Step through a cycle of a TDP Block RAM where only one clock is active. Like
+-- 'writeRam', it accounts for 'Clash.XException.XException' in all values
+-- supplied by the user of the block RAM.
+cycleOne ::
+  forall nAddrs a writeEnable.
+  NFDataX a =>
+  SNat nAddrs ->
+  -- | Determine whether a write enable is active
+  (MaybeX writeEnable -> MaybeX Bool) ->
+  -- | Merge global enable with write enable
+  (MaybeX Bool -> MaybeX writeEnable -> MaybeX writeEnable) ->
+  -- | Update memory with a defined address
+  (Int -> MaybeX writeEnable -> a -> Seq a -> Seq a) ->
+  -- | Previous value
+  a ->
+  -- | Memory
+  Seq a ->
+  -- | Port: enable, address, write enable, write data
+  (MaybeX Bool, MaybeX Int, MaybeX writeEnable, a) ->
+  -- | Updated memory, output value
+  (Seq a, a)
+cycleOne SNat isActiveWe mergeWe updateMem prev ram0 = \case
+  -- RAM is disabled, so we do nothing
+  (IsDefined False, _, _, _) ->
+    (ram0, prev)
+
+  -- RAM is (potentially) enabled, so we run write RAM logic
+  (ena, addr, byteEna0, dat) ->
+    let
+      byteEna1 = mergeWe ena byteEna0
+      (wrote, !ram1) = writeRam (SNat @nAddrs) isActiveWe updateMem addr byteEna1 dat ram0
+
+      out0 = fromMaybe (ram1 `Seq.index` MaybeX.unsafeToA addr) wrote
+      out1 = MaybeX.maybeX deepErrorX (const out0) ena
```

Comment:
The `deepErrorX` will double the `X:` prefix and append a new call stack; see my comment for `unsafeToA`. An alternative is:

```suggestion
      out1 = MaybeX.maybeX (throw . XException) (const out0) ena
```
but that does not use `deepErrorX`. I don't know how to do a `deepErrorX` that doesn't wrap the message with more message. Given that this seems impossible, maybe it's a good idea to just make the wrapping message better than `X: X: Original message`.

```suggestion
      out1 = MaybeX.maybeX enaXmsg (const out0) ena
      enaXmsg msg = deepErrorX $ "Port enable undefined; error message: " <> msg
```